Automate the quarkus superheroes setup (#1)
RedHatPerf/superheroes-automation: README.md

# Superheroes Load Test Automation
Automation for running quarkus superheroes benchmark load tests

## Prerequisites

### QDup
To run these scripts you need [qDup](https://github.com/Hyperfoil/qDup) that you can download directly from the GitHub repository:

```bash
wget https://github.com/Hyperfoil/qDup/releases/download/qDup-0.8.5/qDup-0.8.5-uber.jar

> ```

Reviewer: qDup can be run via jbang as well, it is in the Hyperfoil catalog. You can run it with 

`jbang qdup@hyperfoil http://example.com/scripts/qdup.yml`

Author: You are right, I can change to fully rely on jbang for this automation (I will give it a try first)!

Author: Updated the README

```

* `<native|jvm>`:       which superheroes images you'd like to use, either [`native`](/modes/native.env.yaml) or [`jvm`](/modes/jvm.env.yaml).
* `<benchmark_folder>`: which benchmark you'd like to run among those listed in [/benchmarks](/benchmarks/) folder.
* `[local|remote]`:     where you would like to start the services, either [`local`](/envs/local.env.yaml) or [`remote`](/envs/remote.env.yaml). Default is `local`.
* `[benchmark_params]`: any additional Hyperfoil benchmark template param you want to override, this strictly depends on the HF benchmark definition. Default is empty string.

> [!NOTE]
> If you are willing to use `/envs/remote.env.yaml` be sure to override it in according to your servers

### Examples

#### Get all heroes locally

```bash
./run.sh native get-all-heroes local
```

#### Get all villains locally with 20s duration

```bash
./run.sh native get-all-villains local "-PDURATION=20s"
```

## Run directly using qDup

By default, the `run.sh` will print out the qDup command that it will execute such that you can always 
run the same directly without passing through the script.

```bash

> java -jar <qdup.jar> benchmarks/get-all-heroes/get-all-heroes.env.yaml envs/local.env.yaml modes/native.env.yaml util.yaml hyperfoil.yaml superheroes.yaml qdup.yaml

Reviewer: using jbang this would be;

```suggestion
jbang qdup@hyperfoil benchmarks/get-all-heroes/get-all-heroes.env.yaml envs/local.env.yaml modes/native.env.yaml util.yaml hyperfoil.yaml superheroes.yaml qdup.yaml
```

Author: Correct!

The only change is `qDup@` instead of `qdup@` (note the capital D letter)

> [!NOTE]
> If you want to learn more on Hyperfoil, see the [Hyperfoil website](https://hyperfoil.io).

### JBang

Therefore, the only required tool that you have to install is [jbang](https://www.jbang.dev/documentation/guide/latest). 

Checkout the [installation guide](https://www.jbang.dev/documentation/guide/latest/installation.html) for more details on how you can install it, the suggested approach is by using [sdkman](https://sdkman.io/) - which is installed right away by the qDup script if not present already. 

## Run using script

There is a `run.sh` that aims to makes the superheroes app setup and benchmark execution easier.
It uses qDup under the hood, therefore be sure you have properly installed it in your machine (see [prerequisites](#prerequisites) for more details).

> [!NOTE]
> The script relies on the qDup to be available and its location exported in env variable `QDUP_JAR`

> 

Reviewer: We can remove this as we are using jbang

Author: Good catch, leftover from initial impl.

### Usage

```bash
$ ./run.sh
Usage: ./run.sh <native|jvm> <benchmark_folder> [local|remote] [benchmark_params]
```

> * `<native|jvm>`:       which superheroes images you'd like to use, either [`native`](/modes/native.script.yaml) or [`jvm`](/modes/jvm.script.yaml).

Reviewer: I think we should make it clear here that `Modes` are extensible
```suggestion
* `<native|jvm|custom>`:       which superheroes images you'd like to use, e.g. [`native`](/modes/native.script.yaml) or [`jvm`](/modes/jvm.script.yaml). Modes are extensible by creating a custom (/modes/custom.script.yaml)
```

Author: I agree, I prefer to keep `<custom>` rather than `custom` as the user can provide whatever name they want

* `<benchmark_folder>`: which benchmark you'd like to run among those listed in [/benchmarks](/benchmarks/) folder.

> * `[local|remote]`:     where you would like to start the services, either [`local`](/envs/local.env.yaml) or [`remote`](/envs/remote.env.yaml). Default is `local`.

Reviewer: ```suggestion
* `[local|remote]`:     where you would like to start the services, either [`local`](/envs/local.env.yaml) to run on `localhost` or [`remote`](/envs/remote.env.yaml). Default is `local`.  **Please note:** for `remote` environments, the current user MUST have passwordless ssh access to any remote machines defined in `/envs/remote.env.yaml`
```

Author: Good thing to point out!

* `[benchmark_params]`: any additional Hyperfoil benchmark template param you want to override, this strictly depends on the HF benchmark definition. Default is empty string.

> [!NOTE]

> > If you are willing to use `/envs/remote.env.yaml` be sure to override it in according to your servers

Reviewer: ```suggestion
> If you use `/envs/remote.env.yaml`, please ensure to override variables contained in it with your specific server hostanames
```

### Examples

#### Get all heroes locally

```bash
./run.sh native get-all-heroes local
```

#### Get all villains locally with 20s duration

```bash
./run.sh native get-all-villains local "-PDURATION=20s"
```

## Run directly using qDup

By default, the `run.sh` will print out the qDup command that it will execute such that you can always 
run the same directly without passing through the script.

```bash
jbang qDup@hyperfoil benchmarks/get-all-heroes/get-all-heroes.env.yaml envs/local.env.yaml modes/native.script.yaml util.yaml hyperfoil.yaml superheroes.yaml qdup.yaml
```

Some of those qDup config files are mandatory and cannot be removed:
- `util.yaml`
- `hyperfoil.yaml`
- `superheroes.yaml`
- `qdup.yaml`
- either `envs/local.env.yaml` or `envs/remote.env.yaml`

> - either `modes/native.script.yaml` or `modes/jvm.script.yaml`

Reviewer: this could also be a custom script

- one of `benchmarks/**/*.env.yaml`

## Add more benchmarks

New benchmark scenarios can be added under [/benchmarks](/benchmarks/) folder and the should match the following structure:

```bash
$ tree benchmarks/$BENCHMARK_NAME

benchmarks/get-all-heroes
├── $BENCHMARK_NAME.env.yaml
└── $BENCHMARK_NAME.hf.yaml

1 directory, 2 files
```

As an example:

```bash
$ tree benchmarks/get-all-heroes

benchmarks/get-all-heroes
├── get-all-heroes.env.yaml
└── get-all-heroes.hf.yaml

1 directory, 2 files
```

* `$BENCHMARK_NAME.env.yaml`: contains required qdup states/params to properly retrieve the benchmark definition
* `$BENCHMARK_NAME.hf.yaml`: contains the Hyperfoil benchmark definition

## Custom builds

The automation is generic enough to let you build the superheroes services container images and run them instead of the currently available ones.

### How to build custom images?

> Create a new qDup file under [modes](/modes/) directory, you can take as example the [custom.build.tmpl.yaml](/modes/custom.build.tmpl.yaml) template file using the following format `ENV_IDENTIFIER.env.yaml`.

Reviewer: ```suggestion
Create a new qDup file under [modes](/modes/) directory, you can take as example the [custom.build.tmpl.yaml](/modes/custom.build.tmpl.yaml) template file using the following format `ENV_IDENTIFIER.script.yaml`.
```

Author: thanks, another leftover!
